llvm/lib/Analysis/ScalarEvolution.cpp

```diff
@@ -6019,6 +6019,37 @@ const SCEV *ScalarEvolution::createNodeFromSelectLikePHI(PHINode *PN) {
   return nullptr;
 }
 
+// Returns SCEV for the first operand of a phi if all phi operands have
+// identical opcodes and operands
+// eg.
+// a: %add = %a + %b
+//    br %c
+// b: %add1 = %a + %b
+//    br %c
+// c: %phi = phi [%add, a], [%add1, b]
+// scev(%phi) => scev(%add)
+const SCEV *
+ScalarEvolution::createNodeForPHIWithIdenticalOperands(PHINode *PN) {
+  BinaryOperator *CommonInst = nullptr;
+  for (Value *Incoming : PN->incoming_values()) {
+    BinaryOperator *IncomingInst = dyn_cast<BinaryOperator>(Incoming);
+    if (CommonInst) {
+      if (!(IncomingInst && CommonInst->isIdenticalTo(IncomingInst))) {
```

Comment:
Should we check that all incoming instructions belong to the same loop?

Reply:
If the operands of the phi are `identical`, then the simplification is correct regardless of whether instructions belong to the same loop.

Reply:
Ping?